### PR TITLE
wsgi_logging template configuration

### PR DESCRIPTION
**Pull request by Sakoes**

### CODE REVIEW

File: src/plone/recipe/zope2instance/recipe.py [hunk 1]
```diff
@@ -913,6 +915,8 @@ def build_wsgi_ini(self):
             "profile_flush_at_shutdown": profile_flush_at_shutdown,
             "profile_unwind": profile_unwind,
         }
+
+        # Load wsgi template or use default
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:919]:
```suggestion
        # Load wsgi template from file or use global var as default
```

File: src/plone/recipe/zope2instance/recipe.py [hunk 2]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
             except OSError:
                 raise
 
+        # Load logging template or use default
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:929]:
```suggestion
        # Load wsgi logging template from file or use global var as default
```

File: src/plone/recipe/zope2instance/recipe.py [hunk 3]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
...
         # generate a different [server:main] - useful for Windows
         wsgi_server_main_template = wsgi_server_main_templates.get(
             sys.platform, wsgi_server_main_templates["default"]
         )
         wsgi_options["server_main"] = wsgi_server_main_template % wsgi_options
 
         wsgi_ini = wsgi_ini_template % wsgi_options
+
+        # Parse wsgi.ini
+        configparser.ConfigParser().read_file(io.StringIO(wsgi_ini))
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:951]:
```suggestion
        configparser.ConfigParser().read_string(wsgi_ini)
```

Did you try this ?

File: src/plone/recipe/zope2instance/recipe.py [hunk 4]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
             except OSError:
                 raise
 
+        # Load logging template or use default
+        global logging_ini_template
+        logging_ini_template_path = self.options.get("logging-ini-template")
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:931]:
```suggestion
        wsgi_logging_ini_template_path = self.options.get("wsgi-logging-ini-template")
```

I would be more specific.

File: src/plone/recipe/zope2instance/recipe.py [hunk 5]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
...
+        if logging_ini_template_path:
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:932]:
```suggestion
        if wsgi_logging_ini_template_path:
```

File: src/plone/recipe/zope2instance/recipe.py [hunk 6]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
...
+            try:
+                with open(logging_ini_template_path) as fp:
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:934]:
```suggestion
                with open(wsgi_logging_ini_template_path) as fp:
```

File: src/plone/recipe/zope2instance/recipe.py [hunk 7]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
...
+                    logging_ini_template = fp.read()
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:935]:
```suggestion
                    wsgi_logging_ini_template = fp.read()
```

File: src/plone/recipe/zope2instance/recipe.py [hunk 8]
```diff
@@ -922,13 +926,30 @@ def build_wsgi_ini(self):
...
+            except OSError:
+                raise
+
+        # Add logging config to wsgi.ini
+        wsgi_ini_template += logging_ini_template
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:940]:
```suggestion
        wsgi_ini_template += wsgi_logging_ini_template
```

**Review by gotcha — CHANGES_REQUESTED** [review]
You also need to add a CHANGELOG item.

File: src/plone/recipe/zope2instance/recipe.py (hunk 3 above)
Comment (Sakoes) [reply src/plone/recipe/zope2instance/recipe.py:951]: Tested and implemented with read_string

**Comment by Sakoes** [comment]
Added changes to news/59.feature

File: src/plone/recipe/zope2instance/recipe.py [hunk 9]
```diff
@@ -913,22 +914,48 @@ def build_wsgi_ini(self):
...
+        # Load wsgi logging template from file or use global var as default
+        elif wsgi_logging_ini_template_path:
+            try:
+                with open(wsgi_logging_ini_template_path) as fp:
+                    wsgi_logging_ini_template = fp.read()
+            except OSError:
+                raise
+
+            # Add custom wsgi logging template to wsgi template
+            wsgi_ini_template += wsgi_logging_ini_template
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:943]: Can you beautify this and line 946 that read exactly the same ? (As discussed in flesh.)

**Review by gotcha — CHANGES_REQUESTED** [review]
Can you add the new config in the README ?

Did you check you did not break the test suite ?

File: README.rst [hunk 10]
```diff
@@ -425,6 +425,22 @@ event-log-args
 event-log-kwargs
     Same as `access-log-kwargs` but for the configuration of the event log of Plone.
 
+wsgi-logging-ini-template
+  By default `plone.recipe.zope2instances` uses a hard-coded logging template for the
```
Comment (gotcha) [inline README.rst:429]:
```suggestion
  By default `plone.recipe.zope2instance` uses a hard-coded logging template for the
```

Typo

File: src/plone/recipe/zope2instance/recipe.py [hunk 11]
```diff
@@ -913,22 +914,48 @@ def build_wsgi_ini(self):
...
+
         # generate a different [server:main] - useful for Windows
         wsgi_server_main_template = wsgi_server_main_templates.get(
             sys.platform, wsgi_server_main_templates["default"]
         )
         wsgi_options["server_main"] = wsgi_server_main_template % wsgi_options
 
         wsgi_ini = wsgi_ini_template % wsgi_options
+
+        # Parse wsgi.ini
```
Comment (gotcha) [inline src/plone/recipe/zope2instance/recipe.py:956]:
```suggestion
        # Catch errors in generated wsgi.ini by parsing it before writing the file.```

File: README.rst [hunk 12]
```diff
@@ -425,6 +425,22 @@ event-log-args
...
+  file.
+
+  You can override the template with a custom logging.ini file using this option.
+  All other default wsgi options will be untouched.
+
+  Example::
+
+      wsgi-logging-ini-template = /path/to/logging.ini
+
+  This option cannot be used togehter with the wsgi-ini-template option (see Advanced options: wsgi-ini-template)
```
Comment (gotcha) [inline README.rst:442]:
```suggestion
  This option cannot be used together with the wsgi-ini-template option (see Advanced options: wsgi-ini-template)
```

**Review by gotcha — CHANGES_REQUESTED** [review]
Mostly cosmetics

**Review by gotcha — APPROVED** [review]

**Comment by petschki** [comment]
Hi, tests buildout.coredev are broken since this change ... see https://github.com/plone/buildout.coredev/actions/runs/10943504713/job/30383093298 

```
configparser.DuplicateSectionError: While reading from '<string>' [line 102]: section 'loggers' already exists
```

do we have to change something there or is this change causing it?

File: src/plone/recipe/zope2instance/recipe.py [hunk 13]
```diff
@@ -913,22 +914,51 @@ def build_wsgi_ini(self):
...
+                with open(wsgi_logging_ini_template_path) as fp:
+                    # Add custom wsgi logging template to wsgi template
+                    wsgi_ini_template += fp.read()
+            except OSError:
+                raise
+
+        # Load default global wsgi and logging template
+        else:
+            global wsgi_logging_ini_template
+            wsgi_ini_template += wsgi_logging_ini_template
```
Comment (petschki) [inline src/plone/recipe/zope2instance/recipe.py:949]: This is a problem if you have multiple buildout parts with this recipe. the template string gets multiplied every time the recipe is called. I'll prep a fix for that.

**Pull request merged**